File: python/shield/stream.py
"""
Shield Stream - Streaming encryption for large files.

Processes data in chunks with constant memory usage.
Each chunk is independently authenticated.

Example:
    >>> from shield.stream import StreamCipher
    >>> cipher = StreamCipher.from_password("password", b"salt")
    >>> cipher.encrypt_file("large.bin", "large.bin.enc")
    >>> cipher.decrypt_file("large.bin.enc", "large.bin.dec")
"""

import os
import hmac
import hashlib
import struct
from typing import Iterator, Optional

# Default chunk size: 64KB
DEFAULT_CHUNK_SIZE = 64 * 1024


def _derive_key(password: str, salt: bytes, iterations: int = 100_000) -> bytes:
    """Derive key from password using PBKDF2."""
    return hashlib.pbkdf2_hmac("sha256", password.encode(), salt, iterations)


def _generate_keystream(key: bytes, nonce: bytes, length: int) -> bytes:
    """Generate keystream using SHA256."""
    keystream = b""
    for i in range((length + 31) // 32):
        keystream += hashlib.sha256(key + nonce + struct.pack("<I", i)).digest()
    return keystream[:length]


def _encrypt_block(key: bytes, data: bytes) -> bytes:
    """Encrypt a single block with authentication."""
    nonce = os.urandom(16)
    keystream = _generate_keystream(key, nonce, len(data))
    ciphertext = bytes(a ^ b for a, b in zip(data, keystream))
    mac = hmac.new(key, nonce + ciphertext, hashlib.sha256).digest()[:16]
    return nonce + ciphertext + mac


def _decrypt_block(key: bytes, encrypted: bytes) -> Optional[bytes]:
    """Decrypt a single block with verification."""
    if len(encrypted) < 32:
        return None
    nonce = encrypted[:16]
    ciphertext = encrypted[16:-16]
    mac = encrypted[-16:]

    expected_mac = hmac.new(key, nonce + ciphertext, hashlib.sha256).digest()[:16]
    if not hmac.compare_digest(mac, expected_mac):
        return None

    keystream = _generate_keystream(key, nonce, len(ciphertext))
    return bytes(a ^ b for a, b in zip(ciphertext, keystream))


class StreamCipher:
    """
    Streaming encryption for large files.

    Processes data in chunks with constant memory usage.
    Each chunk is independently authenticated, allowing:
    - Early detection of tampering
    - Constant memory regardless of file size
    - Potential for parallel processing
    """

    def __init__(self, key: bytes, chunk_size: int = DEFAULT_CHUNK_SIZE):
        """
        Initialize with encryption key.

        Args:
            key: 32-byte symmetric key
            chunk_size: Size of each chunk (default: 64KB)
        """
        self.key = key
        self.chunk_size = chunk_size

    @classmethod
    def from_password(
        cls,
        password: str,
        salt: bytes,
        chunk_size: int = DEFAULT_CHUNK_SIZE,
    ) -> "StreamCipher":
        """
        Create StreamCipher from password.

        Args:
            password: User's password
            salt: Salt for key derivation
            chunk_size: Size of each chunk (default: 64KB)

        Returns:
            StreamCipher instance
        """
        key = _derive_key(password, salt)
        return cls(key, chunk_size)

    def encrypt_stream(self, data_iter: Iterator[bytes]) -> Iterator[bytes]:
        """
        Encrypt streaming data.

        Args:
            data_iter: Iterator yielding data chunks

        Yields:
            Encrypted chunks

        Example:
            >>> cipher = StreamCipher(key)
            >>> with open('in.bin', 'rb') as f:
            ...     chunks = iter(lambda: f.read(64*1024), b'')
            ...     for enc in cipher.encrypt_stream(chunks):
            ...         out.write(enc)
        """
        # Header: chunk_size(4) || stream_salt(16)
        stream_salt = os.urandom(16)
        yield struct.pack("<I", self.chunk_size) + stream_salt

        chunk_num = 0
        for data in data_iter:
            if not data:
                break

            # Derive per-chunk key
            chunk_key = hashlib.sha256(
                self.key + stream_salt + struct.pack("<Q", chunk_num)
            ).digest()

            # Encrypt chunk
            encrypted = _encrypt_block(chunk_key, data)

            # Prepend length
            yield struct.pack("<I", len(encrypted)) + encrypted
            chunk_num += 1

        # End marker
        yield struct.pack("<I", 0)

    def decrypt_stream(self, enc_iter: Iterator[bytes]) -> Iterator[bytes]:
        """
        Decrypt streaming data.

        Args:
            enc_iter: Iterator yielding encrypted chunks

        Yields:
            Decrypted chunks

        Raises:
            ValueError: If authentication fails
        """
        # Read header
        header = next(enc_iter)
        _chunk_size = struct.unpack("<I", header[:4])[0]
        stream_salt = header[4:20]

        chunk_num = 0
        buffer = b""

        for data in enc_iter:
            buffer += data

            while len(buffer) >= 4:
                enc_len = struct.unpack("<I", buffer[:4])[0]
                if enc_len == 0:  # End marker
                    return

                if len(buffer) < 4 + enc_len:
                    break

                encrypted = buffer[4 : 4 + enc_len]
                buffer = buffer[4 + enc_len :]

                # Derive per-chunk key
                chunk_key = hashlib.sha256(
                    self.key + stream_salt + struct.pack("<Q", chunk_num)
                ).digest()

                # Decrypt
                decrypted = _decrypt_block(chunk_key, encrypted)
                if decrypted is None:
                    raise ValueError(f"Chunk {chunk_num} authentication failed")

                yield decrypted
                chunk_num += 1

    def encrypt_file(self, in_path: str, out_path: str) -> None:
        """
        Encrypt a file.

        Args:
            in_path: Path to input file
            out_path: Path to output file
        """
        with open(in_path, "rb") as fin, open(out_path, "wb") as fout:
            chunks = iter(lambda: fin.read(self.chunk_size), b"")
            for encrypted in self.encrypt_stream(chunks):
                fout.write(encrypted)

    def decrypt_file(self, in_path: str, out_path: str) -> None:
        """
        Decrypt a file.

        Args:
            in_path: Path to encrypted file
            out_path: Path to output file

        Raises:
            ValueError: If authentication fails
        """

        def reader():
            with open(in_path, "rb") as f:
                # First yield the header (20 bytes: 4 chunk_size + 16 salt)
                header = f.read(20)
                if header:
                    yield header
                # Then yield rest of file in chunks
                while True:
                    data = f.read(self.chunk_size + 48)  # chunk + overhead
                    if not data:
                        break
                    yield data

        with open(out_path, "wb") as fout:
            for decrypted in self.decrypt_stream(reader()):
                fout.write(decrypted)

    def encrypt(self, data: bytes) -> bytes:
        """
        Encrypt data in memory (convenience method).

        Args:
            data: Data to encrypt

        Returns:
            Encrypted data
        """
        result = b""
        for chunk in self.encrypt_stream(iter([data])):
            result += chunk
        return result

    def decrypt(self, encrypted: bytes) -> bytes:
        """
        Decrypt data in memory (convenience method).

        Args:
            encrypted: Encrypted data

        Returns:
            Decrypted data
        """
        # Split header from rest for decrypt_stream which expects header first
        header = encrypted[:20]  # 4 bytes chunk_size + 16 bytes salt
        rest = encrypted[20:]

        result = b""
        for chunk in self.decrypt_stream(iter([header, rest])):
            result += chunk
        return result

File: go/shield/stream.go
package shield

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/binary"
	"errors"
	"io"

	"golang.org/x/crypto/pbkdf2"
)

const (
	// DefaultChunkSize is the default streaming chunk size.
	DefaultChunkSize = 64 * 1024

	// streamSaltSize is the size of the per-stream salt in the header.
	streamSaltSize = 16
	// eofTagSize is the size of the authenticated end-of-stream tag.
	eofTagSize = 32
	// eofLabel is the domain-separation label for the end-of-stream key.
	eofLabel = "shield-stream-eof"
)

// ErrStreamTruncated indicates a stream that is missing its authenticated
// end-of-stream marker (the trailing chunks and/or the end-of-stream tag were
// dropped), which signals truncation.
var ErrStreamTruncated = errors.New("shield: stream truncated: missing end-of-stream marker")

// StreamCipher provides streaming encryption for large data.
type StreamCipher struct {
	key       [KeySize]byte
	chunkSize int
}

// NewStreamCipher creates a new StreamCipher with the given key.
func NewStreamCipher(key []byte, chunkSize int) (*StreamCipher, error) {
	if len(key) != KeySize {
		return nil, ErrInvalidKeySize
	}
	if chunkSize <= 0 {
		chunkSize = DefaultChunkSize
	}
	sc := &StreamCipher{chunkSize: chunkSize}
	copy(sc.key[:], key)
	return sc, nil
}

// StreamCipherFromPassword creates a StreamCipher from password and salt.
func StreamCipherFromPassword(password string, salt []byte, chunkSize int) *StreamCipher {
	key := pbkdf2.Key([]byte(password), salt, Iterations, KeySize, sha256.New)
	if chunkSize <= 0 {
		chunkSize = DefaultChunkSize
	}
	sc := &StreamCipher{chunkSize: chunkSize}
	copy(sc.key[:], key)
	return sc
}

// Encrypt encrypts data in chunks.
//
// Wire format:
//
//	header: chunk_size(u32 LE, 4) || stream_salt(16)
//	frame:  chunk_len(u32 LE, 4) || nonce(16) || ciphertext || mac(16)   (repeated)
//	trailer: u32 LE 0 || eof_tag(32)
//
// The trailer is an authenticated, length-committing end-of-stream marker, so a
// stream truncated at a chunk boundary (with or without a re-appended bare zero
// marker) is detected on decrypt.
func (sc *StreamCipher) Encrypt(plaintext []byte) ([]byte, error) {
	// Header: chunk_size(4 LE) || stream_salt(16)
	streamSalt := make([]byte, streamSaltSize)
	if _, err := rand.Read(streamSalt); err != nil {
		return nil, err
	}

	result := make([]byte, 0, 20+len(plaintext))
	header := make([]byte, 20)
	binary.LittleEndian.PutUint32(header[:4], uint32(sc.chunkSize))
	copy(header[4:], streamSalt)
	result = append(result, header...)

	var chunkNum uint64
	for i := 0; i < len(plaintext); i += sc.chunkSize {
		end := i + sc.chunkSize
		if end > len(plaintext) {
			end = len(plaintext)
		}
		chunk := plaintext[i:end]

		chunkKey := sc.deriveChunkKey(streamSalt, chunkNum)
		encryptedChunk, err := encryptChunk(chunkKey, chunk)
		if err != nil {
			return nil, err
		}

		// Length prefix + encrypted chunk
		lenBuf := make([]byte, 4)
		binary.LittleEndian.PutUint32(lenBuf, uint32(len(encryptedChunk)))
		result = append(result, lenBuf...)
		result = append(result, encryptedChunk...)
		chunkNum++
	}

	// Authenticated end-of-stream trailer: zero marker || eof_tag.
	eofTag := computeEofTag(sc.key[:], streamSalt, chunkNum)
	result = append(result, 0, 0, 0, 0)
	result = append(result, eofTag...)

	return result, nil
}

// Decrypt decrypts chunked data, requiring the authenticated end-of-stream tag.
func (sc *StreamCipher) Decrypt(ciphertext []byte) ([]byte, error) {
	if len(ciphertext) < 20 {
		return nil, ErrCiphertextTooShort
	}

	streamSalt := ciphertext[4:20]
	offset := 20
	var chunkNum uint64
	sawEndMarker := false

	var result []byte
	for offset+4 <= len(ciphertext) {
		chunkLen := binary.LittleEndian.Uint32(ciphertext[offset : offset+4])

		if chunkLen == 0 {
			// Authenticated end-of-stream marker: require the 32-byte tag.
			if offset+4+eofTagSize > len(ciphertext) {
				return nil, ErrStreamTruncated
			}
			tag := ciphertext[offset+4 : offset+4+eofTagSize]
			expected := computeEofTag(sc.key[:], streamSalt, chunkNum)
			if subtle.ConstantTimeCompare(tag, expected) != 1 {
				return nil, ErrAuthenticationFailed
			}
			sawEndMarker = true
			break
		}

		offset += 4
		if offset+int(chunkLen) > len(ciphertext) {
			return nil, ErrCiphertextTooShort
		}

		chunk := ciphertext[offset : offset+int(chunkLen)]
		offset += int(chunkLen)

		chunkKey := sc.deriveChunkKey(streamSalt, chunkNum)
		decrypted, err := decryptChunk(chunkKey, chunk)
		if err != nil {
			return nil, err
		}
		result = append(result, decrypted...)
		chunkNum++
	}

	// A stream that ends without the authenticated marker has been truncated.
	if !sawEndMarker {
		return nil, ErrStreamTruncated
	}

	return result, nil
}

// deriveChunkKey derives the per-chunk key from the master key, stream salt and
// chunk number: SHA256(key || stream_salt || chunk_num as u64 LE).
func (sc *StreamCipher) deriveChunkKey(streamSalt []byte, chunkNum uint64) []byte {
	h := sha256.New()
	h.Write(sc.key[:])
	h.Write(streamSalt)
	var numBuf [8]byte
	binary.LittleEndian.PutUint64(numBuf[:], chunkNum)
	h.Write(numBuf[:])
	return h.Sum(nil)
}

// computeEofTag computes the authenticated end-of-stream tag.
//
//	eof_key = HMAC_SHA256(master_key, "shield-stream-eof")
//	eof_tag = HMAC_SHA256(eof_key, stream_salt || chunk_count as u64 LE)
//
// The tag commits to the stream salt and total chunk count (a length
// commitment); an attacker cannot forge a matching tag without the master key,
// so truncation -- including re-appending a bare zero marker -- is detected.
func computeEofTag(masterKey, streamSalt []byte, chunkCount uint64) []byte {
	ek := hmac.New(sha256.New, masterKey)
	ek.Write([]byte(eofLabel))
	eofKey := ek.Sum(nil)

	t := hmac.New(sha256.New, eofKey)
	t.Write(streamSalt)
	var countBuf [8]byte
	binary.LittleEndian.PutUint64(countBuf[:], chunkCount)
	t.Write(countBuf[:])
	return t.Sum(nil)
}

func encryptChunk(key, data []byte) ([]byte, error) {
	nonce := make([]byte, NonceSize)
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}

	keystream := generateKeystream(key, nonce, len(data))
	ciphertext := make([]byte, len(data))
	for i := range data {
		ciphertext[i] = data[i] ^ keystream[i]
	}

	mac := hmac.New(sha256.New, key)
	mac.Write(nonce)
	mac.Write(ciphertext)
	tag := mac.Sum(nil)[:MACSize]

	result := make([]byte, NonceSize+len(ciphertext)+MACSize)
	copy(result[:NonceSize], nonce)
	copy(result[NonceSize:NonceSize+len(ciphertext)], ciphertext)
	copy(result[NonceSize+len(ciphertext):], tag)

	return result, nil
}

func decryptChunk(key, encrypted []byte) ([]byte, error) {
	if len(encrypted) < NonceSize+MACSize {
		return nil, ErrCiphertextTooShort
	}

	nonce := encrypted[:NonceSize]
	ciphertext := encrypted[NonceSize : len(encrypted)-MACSize]
	receivedMAC := encrypted[len(encrypted)-MACSize:]

	mac := hmac.New(sha256.New, key)
	mac.Write(nonce)
	mac.Write(ciphertext)
	expectedMAC := mac.Sum(nil)[:MACSize]

	if subtle.ConstantTimeCompare(receivedMAC, expectedMAC) != 1 {
		return nil, ErrAuthenticationFailed
	}

	keystream := generateKeystream(key, nonce, len(ciphertext))
	plaintext := make([]byte, len(ciphertext))
	for i := range ciphertext {
		plaintext[i] = ciphertext[i] ^ keystream[i]
	}

	return plaintext, nil
}

// EncryptReader encrypts data from a reader to a writer.
func (sc *StreamCipher) EncryptReader(r io.Reader, w io.Writer) error {
	data, err := io.ReadAll(r)
	if err != nil {
		return err
	}
	encrypted, err := sc.Encrypt(data)
	if err != nil {
		return err
	}
	_, err = w.Write(encrypted)
	return err
}

// DecryptReader decrypts data from a reader to a writer.
func (sc *StreamCipher) DecryptReader(r io.Reader, w io.Writer) error {
	data, err := io.ReadAll(r)
	if err != nil {
		return err
	}
	decrypted, err := sc.Decrypt(data)
	if err != nil {
		return err
	}
	_, err = w.Write(decrypted)
	return err
}
